sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -176,10 +176,16 @@ object TypeCoercion {
   }
 
   private def findWiderCommonType(types: Seq[DataType]): Option[DataType] = {
-    types.foldLeft[Option[DataType]](Some(NullType))((r, c) => r match {
-      case Some(d) => findWiderTypeForTwo(d, c)
-      case None => None
-    })
+    // findWiderTypeForTwo doesn't satisfy the associative law, i.e. (a op b) op c may not equal
+    // to a op (b op c). This is only a problem for StringType. Excluding StringType,
+    // findWiderTypeForTwo satisfies the associative law. For instance, (TimestampType,
+    // IntegerType, StringType) should have StringType as the wider common type.
+    val (stringTypes, nonStringTypes) = types.partition(_ == StringType)
```

Comment:
Out of curiosity, does this work with array types too (array of string vs array of non string types)?

Reply:
It's expected to , let me also fix it for array types. Thanks!